Prevent Yarr::Interpreter's evaluation stack from growing unboundedly.

MenloDorian opened the pull request:
#### 424c8d8832690078419ba0586eb81416d7f16f9d
<pre>
Prevent Yarr::Interpreter&apos;s evaluation stack from growing unboundedly.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287074">https://bugs.webkit.org/show_bug.cgi?id=287074</a>
<a href="https://rdar.apple.com/143786123">rdar://143786123</a>

Reviewed by Keith Miller and Michael Saboff.

Currently, Yarr::Interpreter&apos;s evaluation stack (see BytecodePattern::m_allocator) is allowed
to grow unboundedly until we exhaust all memory.  We should bound it instead to a max capacity
limit.

1. The evaluation stack uses the BumpPointerAllocator class.  We enhanced BumpPointerAllocator&apos;s
   startAllocator() to take a maxCapacity value.  This maxCapacity value is used internally
   to compute a remainingCapacity value.

   The BumpPointerAllocator works by creating a link list of BumpPointerPools.  Each BumpPointerPool
   will now track the remainingCapacity should it needs to allocate the next BumpPointerPool in
   the link.  The size of the current BumpPointerPool will be deducted from remainingCapacity.
   When requested growth exceeds the remainingCapacity, BumpPointerPool::create() will fail to
   create another pool.

2. Introduced JSC::Options::maxRegExpStackSize() to define the max capacity.  The current default
   for this option value is 4M.

3. Yarr::interpret() has always returned offsetNoMatch if any errors occur during evaluation.
   When we encounter this new error condition where we&apos;ve exhausted the evaluation stack, we&apos;ll
   do the same thing.

4. Fixed a potential memory leak in freeParenthesesDisjunctionContext().  It was freeing the
   ParenthesesDisjunctionContext memory without calling its destructor.  This used to be fine
   because ParenthesesDisjunctionContext used to be trivially destructible.  That is no longer
   the case since a BitVector field got added into it.  Hence, we fixed
   freeParenthesesDisjunctionContext() to also call ~ParenthesesDisjunctionContext() before
   we dealloc its memory.

5. Also removed 2 unused variants of the Yarr::interpret() method.

* JSTests/stress/regexp-huge-oom.js:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
(JSC::Yarr::errorMessage):
* Source/JavaScriptCore/yarr/YarrErrorCode.h:
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::allocDisjunctionContext):
(JSC::Yarr::Interpreter::freeDisjunctionContext):
(JSC::Yarr::Interpreter::allocParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::freeParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::matchParentheses):
(JSC::Yarr::Interpreter::backtrackParentheses):
(JSC::Yarr::interpret):
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
* Source/WTF/wtf/BumpPointerAllocator.h:
(WTF::BumpPointerPool::BumpPointerPool):
(WTF::BumpPointerPool::create):
(WTF::BumpPointerPool::ensureCapacityCrossPool):
(WTF::BumpPointerAllocator::startAllocator):
* Tools/TestWebKitAPI/Tests/WTF/BumpPointerAllocator.cpp:
(TestWebKitAPI::TEST(WTF_BumpPointerAllocator, AllocationWithOnlySmallerPoolsAvailable)):

Canonical link: <a href="https://commits.webkit.org/290198@main">https://commits.webkit.org/290198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61d12b6f69e0154477a23bc7f9993184e75ae5e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89225 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94207 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91276 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9137 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16974 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68744 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26419 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92227 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7007 "Found 1 new test failure: fast/forms/ios/file-upload-panel.html (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49105 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6756 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39093 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82024 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77117 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96040 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88001 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16405 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77622 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16661 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76916 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21331 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9523 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13991 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16419 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21730 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110494 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16160 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26513 "Found 9 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.bytecode-cache, microbenchmarks/memcpy-wasm-small.js.bytecode-cache, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-collect-continuously, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19611 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Tools/TestWebKitAPI/Tests/WTF/BumpPointerAllocator.cpp:
```diff
@@ -30,7 +31,7 @@ namespace TestWebKitAPI {
 {
     WTF::BumpPointerAllocator allocator;
-    WTF::BumpPointerPool* pool = allocator.startAllocator();
+    WTF::BumpPointerPool* pool = allocator.startAllocator(4 * 1024 * 1024);
```
Nit: I would do `4 * MB`

MenloDorian replied:
Fixed.

kmiller68 commented on Source/WTF/wtf/BumpPointerAllocator.h:
```diff
@@ -98,16 +98,17 @@ class BumpPointerPool {
     }
 
-    static BumpPointerPool* create(size_t minimumCapacity = 0)
+    static BumpPointerPool* create(size_t remainingCapacity, size_t minimumCapacity = 0)
```
Isn't this really `maximumCapacity` within the context of this function? Probably also worth an `ASSERT(minimumCapacity <= maximumCapacity)`

MenloDorian replied:
Keith commented elsewhere that this is due to a misunderstanding.

kmiller68 commented on Source/WTF/wtf/BumpPointerAllocator.h:
```diff
@@ -123,9 +124,12 @@ class BumpPointerPool {
                 return nullptr;
         }
 
+        if (poolSize > remainingCapacity)
+            return nullptr;
```
If the parameter is actually "remaining capacity" this check isn't needed. If it's actually a maximum why do we need to compute `poolSize` above? I think you could basically just do `PageAlloaction::allocate(roundUpToMultipleOf(maximumCapacity, std::max(static_cast<size_t>(MINIMUM_BUMP_POOL_SIZE), WTF::pageSize()))`. I also don't know what `minimumCapacity` is doing at that point either, FWIW.

MenloDorian replied:
`minimumCapacity` is used for cases where the client wants to specify a minimum size to allocate for the pool.  Otherwise, the allocated pool (for growing the eval stack) might not fit the allocation request.

`remainingCapacity` is an accurate description for this pool allocation.  If we rename it to `maximumCapacity`, then the way to think of it is as the max capacity for remaining allocation.  Hence, `remainingCapacity` is the better name.

> If the parameter is actually "remaining capacity" this check isn't needed.
Why is this check not needed?  The point is that `PageAllocation::allocate` may succeed, but we don't want to allow it because we're bounding the memory use of the eval stack (as implemented by the chain of `BumpPointerPool`s.  Can you explain what you mean here?

kmiller68 replied:
I agree once it's passed to the `BumpPointerPool` constructor it's the remaining capacity but from the caller of this function's perspective it's the total capacity since the "current capacity" is 0. If I saw the signature `BumpPointerPool* BumpPointerPool::create(size_t remainingCapacity, size_t minimumCapacity = 0)` first I would be confused about what it means for something to "remain" for a thing that doesn't exist yet. Then in a void I would expect that the returned pool would happily allocate until I hit `remainingCapacity + minimumCapacity` but I actually get a pool that will allocate until `remainingCapacity` bytes are used. I don't think any such confusion exists if the function is `BumpPointerPool* BumpPointerPool::create(size_t maximumCapacity, size_t minimumCapacity = 0)`

kmiller68 replied:
Oh I misunderstood this class. I thought this create was a public API not an internal one. Let me look again.

dhecht commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -237,13 +252,20 @@ class Interpreter {
+#if ASSERT_ENABLED
+        ASSERT(context->m_magicNumber == ParenthesesDisjunctionContext::magicNumber);
+        context->m_magicNumber = 0;
+#endif
```
what class of bugs are the magicNumbers meant to catch? I think it will only catch double-free as written?

Given that the bug here was the missing destructor call, maybe move this code inside the destructor itself, and then right before the `dealloc` call, assert that the magic number is 0.  Still, I'm not sure how much utility adding these magicNumbers provides given that removing the call to the destructor erroneously seems unlikely now that the call exists.

MenloDorian replied:
The BumpPointerAllocator is meant to work like a stack: allocs / deallocs work like push / pops.  However, it is being used to alloc memory for objects that we `new` and destruct.  The correctness of the stack ordering depends entirely on client code i.e. the Yarr::Interpreter code in this case.  If a bug creeps into the ordering, we may end up dealloc'ing the wrong piece of memory.

The magicNumbers are meant as a sanity check to help catch is/when we introduce such bugs.

msaboff commented on Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
```diff
@@ -65,6 +66,7 @@ ASCIILiteral errorMessage(ErrorCode error)
         REGEXP_ERROR_PREFIX "invalid property expression"_s,                          // InvalidUnicodePropertyExpression
         REGEXP_ERROR_PREFIX "too many nested disjunctions"_s,                         // TooManyDisjunctions
         REGEXP_ERROR_PREFIX "pattern exceeds string length limits"_s,                 // OffsetTooLarge
+        "Exhausted regular expression stack"_s,                                       // OutOfMemory
```
The existing list is all syntax errors, either discovered during parsing or while converting to JIT code or ByteCompiling in the interpreter.

Please move this string down to the end and add a comment that this is a runtime error.  You'll need to update the enum as well.

MenloDorian replied:
Fixed.  I've removed this `ErrorCode::OutOfMemory`.  Turns out to not be needed.

msaboff commented on Source/JavaScriptCore/yarr/YarrErrorCode.h:
```diff
@@ -64,6 +65,7 @@ enum class ErrorCode : uint8_t {
     InvalidUnicodePropertyExpression,
     TooManyDisjunctions,
     OffsetTooLarge,
+    OutOfMemory,
```
I think it makes sense to add this error to hasHardError() below as another error that is not a hard (syntax) error.

MenloDorian replied:
Fixed.  I've updated the enum list to be clearer.  But in the end, `OutOfMemory` was removed since it's no longer needed.

msaboff reviewed (approved):
r=me with comments.

kmiller68 reviewed (approved):
r=me too.

Commit:
Prevent Yarr::Interpreter's evaluation stack from growing unboundedly.
https://bugs.webkit.org/show_bug.cgi?id=287074
rdar://143786123

Reviewed by Keith Miller and Michael Saboff.

Currently, Yarr::Interpreter's evaluation stack (see BytecodePattern::m_allocator) is allowed
to grow unboundedly until we exhaust all memory.  We should bound it instead to a max capacity
limit.

1. The evaluation stack uses the BumpPointerAllocator class.  We enhanced BumpPointerAllocator's
   startAllocator() to take a maxCapacity value.  This maxCapacity value is used internally
   to compute a remainingCapacity value.

   The BumpPointerAllocator works by creating a link list of BumpPointerPools.  Each BumpPointerPool
   will now track the remainingCapacity should it needs to allocate the next BumpPointerPool in
   the link.  The size of the current BumpPointerPool will be deducted from remainingCapacity.
   When requested growth exceeds the remainingCapacity, BumpPointerPool::create() will fail to
   create another pool.

2. Introduced JSC::Options::maxRegExpStackSize() to define the max capacity.  The current default
   for this option value is 4M.

3. Yarr::interpret() has always returned offsetNoMatch if any errors occur during evaluation.
   When we encounter this new error condition where we've exhausted the evaluation stack, we'll
   do the same thing.

4. Fixed a potential memory leak in freeParenthesesDisjunctionContext().  It was freeing the
   ParenthesesDisjunctionContext memory without calling its destructor.  This used to be fine
   because ParenthesesDisjunctionContext used to be trivially destructible.  That is no longer
   the case since a BitVector field got added into it.  Hence, we fixed
   freeParenthesesDisjunctionContext() to also call ~ParenthesesDisjunctionContext() before
   we dealloc its memory.

5. Also removed 2 unused variants of the Yarr::interpret() method.

* JSTests/stress/regexp-huge-oom.js:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/yarr/YarrErrorCode.cpp:
(JSC::Yarr::errorMessage):
* Source/JavaScriptCore/yarr/YarrErrorCode.h:
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::allocDisjunctionContext):
(JSC::Yarr::Interpreter::freeDisjunctionContext):
(JSC::Yarr::Interpreter::allocParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::freeParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::matchParentheses):
(JSC::Yarr::Interpreter::backtrackParentheses):
(JSC::Yarr::interpret):
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
* Source/WTF/wtf/BumpPointerAllocator.h:
(WTF::BumpPointerPool::BumpPointerPool):
(WTF::BumpPointerPool::create):
(WTF::BumpPointerPool::ensureCapacityCrossPool):
(WTF::BumpPointerAllocator::startAllocator):
* Tools/TestWebKitAPI/Tests/WTF/BumpPointerAllocator.cpp:
(TestWebKitAPI::TEST(WTF_BumpPointerAllocator, AllocationWithOnlySmallerPoolsAvailable)):

Canonical link: https://commits.webkit.org/290198@main
